source/common/network/win32_redirect_records_option_impl.h

```diff
@@ -11,22 +11,24 @@
 namespace Envoy {
 namespace Network {
 
-#define IOCTL_LEVEL 1
+// `IOCTL` controls unlike socket options do not have level paremeter. So we arbitrarily define one
+// in Envoy.
+#define ENVOY_IOCTL_LEVEL 1
 
 #ifdef SIO_SET_WFP_CONNECTION_REDIRECT_RECORDS
 #define ENVOY_SOCKET_REDIRECT_RECORDS                                                              \
-  ENVOY_MAKE_SOCKET_OPTION_NAME(IOCTL_LEVEL, SIO_SET_WFP_CONNECTION_REDIRECT_RECORDS)
+  ENVOY_MAKE_SOCKET_OPTION_NAME(ENVOY_IOCTL_LEVEL, SIO_SET_WFP_CONNECTION_REDIRECT_RECORDS)
 #else
 #define ENVOY_SOCKET_REDIRECT_RECORDS Network::SocketOptionName()
 #endif
 
-class IoctlSocketOptionImpl : public Socket::Option, Logger::Loggable<Logger::Id::connection> {
+class Win32RedirectRecordsOptionImpl : public Socket::Option,
+                                       Logger::Loggable<Logger::Id::connection> {
 public:
-  IoctlSocketOptionImpl(envoy::config::core::v3::SocketOption::SocketState in_state,
-                        Network::SocketOptionName optname, void* inBuffer,
-                        unsigned long inBufferSize, void* outBuffer, unsigned long outBufferSize)
-      : in_state_(in_state), optname_(optname), inBuffer_(inBuffer), inBuffer_size_(inBufferSize),
-        outBuffer_(outBuffer), outBuffer_size_(outBufferSize) {}
+  Win32RedirectRecordsOptionImpl(envoy::config::core::v3::SocketOption::SocketState in_state,
+                                 Network::SocketOptionName optname,
```

Comment:
I think `in_state` and `optname` can be hard-coded into the code for this class, instead of being parameters/member-variables.. Unless you have plans to use this class in another way.

Reply:
`optname ` is needed to check if the option `hasValue()` but the other is not needed. Fixed in 3437dc6078878c8282a466a53d09441d3b8437da

Reply:
I'm pretty sure you can use `ENVOY_SOCKET_REDIRECT_RECORDS` directly, as in `ENVOY_SOCKET_REDIRECT_RECORDS.hasValue()`. The macro evaluates to a `SocketOptionName`, which is either empty or has the correct value.

Reply:
Wouldn't that construct a new instance of `struct SocketOptionName` each time we call `hasValue()`?

Reply:
I think you could have a `static constexpr SocketOptionName` initialized from the macro in the file and use that everytime.

Reply:
I have some concerns with deleting `optname` because of the following:

1. The class is not trivially constructable. which means that this would violate envoy style guide,

> The Google C++ style guide points out that non-PoD static and global variables are forbidden. This includes types such as std::string. We encourage the use of the advice in the C++ FAQ on the static initialization fiasco for how to best handle this.

2. It breaks consistency will all other socket options.


That being said, I see what you are saying. Since all the socket option support info is known at compile time we don't need to initialize it every time we construct a socket option instance. I think this is a problem with how all the socket options behave right now.

Reply:
https://godbolt.org/z/jqTW7T

Reply:
I think it's ok to break consistency with other socket options, given that this one is kind of special (actually ioctl, not sockopt; uses special data). 

You could make one with `CONSTRUCT_ON_FIRST_USE`, or even just make a new instance on the stack everytime you need it; it's construction isn't a performance concern. With what you have now, you're creating a new instance everytime the object is created at the call to the constructor.

Reply:
fixed in https://github.com/envoyproxy/envoy/pull/14753/commits/c2987c9223a969066c85373795ef7dec8ade79c9